File: app/src/main/java/com/lee/leewanandroid/algorithm/java/tree/AVLTree.java
package com.lee.leewanandroid.algorithm.java.tree;

import com.lee.leewanandroid.algorithm.java.tree.node.Node;

/**
 * @Description: AVLTree
 * @Author: lihuayong
 * @CreateDate: 2019-06-24 14:27
 * @UpdateUser:
 * @UpdateDate: 2019-06-24 14:27
 * @UpdateRemark:
 * @Version: 1.0
 */
public class AVLTree<K extends Comparable<K>, V> extends BSTree<K, V> implements OnTreeChangedListener<K, V> {

    public AVLTree() {
        setOnTreeChangedListener(this);
    }

    @Override
    public void fixAfterInsertion(Node<K, V> node) {
        rebuildAfterTreeChanged(node);
    }

    @Override
    public void fixAfterDeletion(Node<K, V> node, boolean delete) {
        if (!delete) {
            //如果没有删除，因为平衡二叉树平衡需要计算节点高度，必须要把该节点删除了。
            //然后开始计算其父节点的平衡性
            if (node.parent != null) {
                if (node == node.parent.left)
                    node.parent.left = null;
                else if (node == node.parent.right)
                    node.parent.right = null;
            }
            rebuildAfterTreeChanged(node.parent);
            node.parent = null;
        } else {
            rebuildAfterTreeChanged(node);
        }
    }

    private void rebuildAfterTreeChanged(Node<K, V> node) {
        Node<K, V> p = node;
        while (p != null) {
            updateNodeHeight(p);
            int bf = p.balance();
            if (bf >= 2) {
                rotationAfterChanged(p, true);
            } else if (bf <= -2) {
                rotationAfterChanged(p, false);
            }
            p = p.parent;
        }
    }

    private void rotationAfterChanged(Node<K, V> p, boolean left) {
        if (left) {
            Node<K, V> lChild = p.left;
            if (lChild.balance() >= 0) { //右旋
                rotateRight(p);
            } else {
                rotateLeft(p.left);
                rotateRight(p);
            }
        } else {
            Node<K, V> rChild = p.right;
            if (rChild.balance() <= 0) { //左旋
                rotateLeft(p);
            } else {
                rotateRight(p.right);
                rotateLeft(p);
            }
        }
    }

    private void rotateLeft(Node<K, V> p) {
        if (p != null) {
            Node<K, V> r = p.right;
            p.right = r.left;
            if (r.left != null)
                r.left.parent = p;
            r.parent = p.parent;
            if (p.parent == null)
                root = r;
            else if (p.parent.left == p)
                p.parent.left = r;
            else
                p.parent.right = r;
            r.left = p;
            p.parent = r;

            // Update heights
            updateNodeHeight(p);
            updateNodeHeight(r);

            println("after rotateLeft " + p.key);
            printTree();
        }
    }

    private void rotateRight(Node<K, V> p) {
        if (p != null) {
            Node<K, V> l = p.left;
            p.left = l.right;
            if (l.right != null) l.right.parent = p;
            l.parent = p.parent;
            if (p.parent == null)
                root = l;
            else if (p.parent.right == p)
                p.parent.right = l;
            else p.parent.left = l;
            l.right = p;
            p.parent = l;

            // Update heights
            updateNodeHeight(p);
            updateNodeHeight(l);

            println("after rotateRight " + p.key);
            printTree();
        }
    }
}

File: app/src/main/java/com/lee/leewanandroid/algorithm/tree/AVLTree.kt
package com.lee.leewanandroid.algorithm.tree

/**
 *
 * @Description:    AVLTree 平衡二叉树 插入删除的非递归实现版本
 * @Author:         lihuayong
 * @CreateDate:     2019-06-18 14:23
 * @UpdateUser:
 * @UpdateDate:     2019-06-18 14:23
 * @UpdateRemark:
 * @Version:        1.0
 *
 *  in order [47 58 62 73 88 93 99 100] -> insert [51]
 *              62
 *            /   \
 *          58     88
 *         /      /  \
 *        47     73  99
 *          \        / \
 *         51      93 100
 *
 */
@Suppress("KDocUnresolvedReference")
class AVLTree : BinarySearchTree<Int>() {

    override fun insert(value: Int): Boolean {
        val insertP = insertInternal(value)
        heightR(root)
        println("after insert $value, parent's value = ${insertP?.value}")
        printTree()
        rebuild(insertP)
        return insertP == null
    }

    private fun rebuild(insertP: TNode<Int>?) {
        //https://www.cnblogs.com/qm-article/p/9349681.html
        var p = insertP
        while (p != null) {
            val lh = p.left.height()
            val rh = p.right.height()
            val bf = lh - rh
            if (bf >= 2) {
                rebuildAfterInsert(p, true)
            } else if (bf <= -2) {
                rebuildAfterInsert(p, false)
            }
            p = p.parent
        }
    }

    private fun rebuildAfterInsert(p: TNode<Int>, left: Boolean) {
        if (left) {
            val lChild = p.left
            if (lChild?.left != null) { //右旋
                rightRotation(p)
            } else if (lChild?.right != null) {
                leftRightRotation(p)
            }
        } else {
            val rChild = p.right
            if (rChild?.right != null) { //左旋
                leftRotation(p)
            } else if (rChild?.left != null) {
                rightLeftRotation(p)
            }
        }
    }

    override fun remove(value: Int): Boolean {
        val removeP = removeInternal(value)
        heightR(root)
        println("after remove $value, parent's value = ${removeP?.value}")
        printTree()
        rebuild(removeP)
        return removeP == null
    }

    /**
     *  \           \
     *   [A]         B
     *     \   =>   / \
     *      B      A   C
     *     / \      \
     *        C
     * 左旋
     */
    private fun leftRotation(nodeA: TNode<Int>): TNode<Int> {
        val nodeB = nodeA.right as TNode<Int>
        val nodeT = nodeB.left as TNode<Int>?
        if (nodeA.parent != null) {
            nodeA.parent?.let { p ->
                if (p.left == nodeA) {
                    p.left = nodeB
                } else {
                    p.right = nodeB
                }
                nodeB.parent = p
            }
        } else {
            //nodeA 是root节点
            nodeB.parent = null
            root = nodeB
        }

        nodeB.left = nodeA
        nodeA.parent = nodeB

        nodeA.right = nodeT
        nodeT?.parent = nodeA

        heightR(root)
        println("\n\nafter left rotation, nodeA value = [${nodeA.value}]")
        printTree()
        return nodeB
    }

    /**
     *
     *     [A]        /
     *     /         B
     *    B    =>   / \
     *   / \       C  A
     *  C  t         /
     *              t
     * 右旋
     */
    private fun rightRotation(nodeA: TNode<Int>): TNode<Int> {
        val nodeB = nodeA.left as TNode<Int>
        val nodeT = nodeB.right as TNode<Int>?
        if (nodeA.parent != null) {
            nodeA.parent?.let { p ->
                if (p.left == nodeA) {
                    p.left = nodeB
                } else {
                    p.right = nodeB
                }
                nodeB.parent = p
            }
        } else {
            //nodeA 是root节点
            nodeB.parent = null
            root = nodeB
        }
        nodeB.right = nodeA
        nodeA.parent = nodeB

        nodeA.left = nodeT
        nodeT?.parent = nodeA

        heightR(root)
        println("\n\nafter right rotation, nodeA value = [${nodeA.value}]")
        printTree()
        return nodeB
    }

    /**
     *     /             /
     *    A            [A]        /
     *   /             /         C
     * [B]      =>    C    =>   / \
     *   \           /         B  A
     *    C         B
     * 先左旋，再右旋
     */
    private fun leftRightRotation(nodeA: TNode<Int>): TNode<Int> {
        val nodeB = nodeA.left
        leftRotation(nodeB as TNode<Int>)
        return rightRotation(nodeA)
    }

    /**
     *  \         \          \
     *   A        [A]         C
     *    \    =>   \   =>   / \
     *    [B]        C      A   B
     *    /           \
     *  C              B
     * 先右旋，再左旋
     */
    private fun rightLeftRotation(nodeA: TNode<Int>): TNode<Int> {
        val nodeB = nodeA.right
        rightRotation(nodeB as TNode<Int>)
        return leftRotation(nodeA)
    }
}